File: react_native_app/src/core/network/apiClient.ts
import axios, { type AxiosInstance } from 'axios';

import { NetworkException, ServerException } from '../error/exceptions';

/**
 * Cliente HTTP base. Encapsula el acceso a la API REST (DummyJSON) y traduce
 * los problemas técnicos a `ServerException` / `NetworkException`.
 */
export class ApiClient {
  private readonly http: AxiosInstance;

  constructor(baseURL = 'https://dummyjson.com') {
    this.http = axios.create({
      baseURL,
      timeout: 15000,
      headers: { 'Content-Type': 'application/json' },
    });
  }

  get<T = unknown>(path: string): Promise<T> {
    return this.send(() => this.http.get<T>(path));
  }

  post<T = unknown>(path: string, body: unknown): Promise<T> {
    return this.send(() => this.http.post<T>(path, body));
  }

  put<T = unknown>(path: string, body: unknown): Promise<T> {
    return this.send(() => this.http.put<T>(path, body));
  }

  delete<T = unknown>(path: string): Promise<T> {
    return this.send(() => this.http.delete<T>(path));
  }

  private async send<T>(request: () => Promise<{ data: T }>): Promise<T> {
    try {
      const response = await request();
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        // Hay respuesta del servidor con código de error (4xx/5xx).
        if (error.response) {
          throw new ServerException(`HTTP ${error.response.status}`);
        }
        // No hubo respuesta: timeout, DNS, sin conexión.
        throw new NetworkException();
      }
      throw new ServerException();
    }
  }
}

File: flutter_app/lib/core/network/api_client.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import '../error/exceptions.dart';

/// Cliente HTTP base. Encapsula el acceso a la API REST (DummyJSON) y traduce
/// los problemas técnicos a [ServerException] / [NetworkException].
class ApiClient {
  final http.Client _client;
  final String baseUrl;

  ApiClient({http.Client? client, this.baseUrl = 'https://dummyjson.com'})
      : _client = client ?? http.Client();

  Future<dynamic> get(String path) => _send(() => _client.get(_uri(path)));

  Future<dynamic> post(String path, Map<String, dynamic> body) =>
      _send(() => _client.post(_uri(path), headers: _headers, body: jsonEncode(body)));

  Future<dynamic> put(String path, Map<String, dynamic> body) =>
      _send(() => _client.put(_uri(path), headers: _headers, body: jsonEncode(body)));

  Future<dynamic> delete(String path) => _send(() => _client.delete(_uri(path)));

  Uri _uri(String path) => Uri.parse('$baseUrl$path');

  Map<String, String> get _headers => const {'Content-Type': 'application/json'};

  Future<dynamic> _send(Future<http.Response> Function() request) async {
    try {
      final response = await request();
      if (response.statusCode >= 200 && response.statusCode < 300) {
        if (response.body.isEmpty) return null;
        return jsonDecode(response.body);
      }
      throw ServerException('HTTP ${response.statusCode}');
    } on SocketException {
      throw NetworkException();
    } on HttpException {
      throw NetworkException();
    } on FormatException {
      throw ServerException('Respuesta inválida del servidor');
    }
  }
}
